// add-remove/index.js
var core = require('grid-core');
var absDivs = require('grid-absolute-divs');
var util = require('grid-util');

module.exports = function (container) {
    var $container = $(container);
    var grid = core(container);
    var cells = [];
    var buffer = 50; //number of cells to render on either side
    var useScroller = true;

    var $scroller;

    function buildCells() {
        $container.empty();
        grid.calc();
        $scroller = $('<div>');
        if (useScroller) {
            $scroller.css({
                height: grid.rowPixelRange.end,
                width: grid.colPixelRange.end
            });
        }
        $container.append($scroller);
        $container.css({'overflow': 'auto'});
    }

    function renderWindow() {
        var rowRange = Number.range(grid.rowOffset - buffer, grid.rowOffset + grid.onScreenRows + buffer);
        var colRange = Number.range(grid.colOffset - buffer, grid.colOffset + grid.onScreenCols + buffer);
        $scroller.remove();
        grid.eachCell(function (r, c) {
            var cell = cells[r] && cells[r][c];
            if (rowRange.contains(r) && colRange.contains(c)) {
                if (!cell) {

                    cell = $('<div>').css({
                        height: grid.cellHeight + 'px',
                        width: grid.cellWidth + 'px',
                        top: r * grid.cellHeight + 'px',
                        left: c * grid.cellWidth + 'px',
                        position: 'absolute'
                    }).addClass(util.getCellClass(r, c));
                    cell[0].innerHTML = r + ', ' + c;
                    cells[r] = cells[r] || [];
                    cells[r][c] = cell;
                }
                $scroller.append(cell);
            } else {
                if (cell) {
                    cell.remove();
                    if (cells[r]) {
                        cells[r][c] = undefined;
                    }
                }
            }
        });
        $container.append($scroller);
    }

    var updateOffsetsAndDraw = function () {
        grid.updateScroll($container.scrollTop(), $container.scrollLeft());
        if (grid.calcOffsets()) {
            renderWindow();
        }
    }.debounce(500);

    $container.on('scroll', function () {
        updateOffsetsAndDraw();
    });

    return Object.merge(grid, {
        rebuild: function () {
            buildCells();
            renderWindow();
        },
        setUseScroller: function (use) {
            useScroller = use;
        }
    });
}
