// render-once/index.js
var mousewheel = require('mousewheel');

module.exports = function (container) {
    var $container = $(container);

    //build grid
    var cellHeight = 30, cellWidth = 100;
    var scrollTop = 0, scrollLeft = 0;
    var rowOffset = 0, colOffset = 0;
    var maxRows = 10000, maxCols = 10000;
    var rows, cols;
    var rowRange;
    var colRange;
    var cells = [];
    var useTextNodes = false;

    function getCellClass(r, c) {
        return 'row-' + r + ' col-' + c + ' cell ' + (r % 2 === 1 ? 'odd' : 'even');
    }

    function buildCells() {
        rows = Math.floor($container.height() / cellHeight), cols = Math.floor($container.width() / cellWidth);
        rowRange = Number.range(0, maxRows * cellHeight);
        colRange = Number.range(0, maxCols * cellWidth);
        $container.empty();
        for (var r = 0; r < rows; r++) {
            cells[r] = [];
            for (var c = 0; c < cols; c++) {
                var cell = $('<div>').css({
                    height: cellHeight + 'px',
                    width: cellWidth + 'px',
                    top: r * cellHeight + 'px',
                    left: c * cellWidth + 'px',
                    position: 'absolute'
                }).addClass(getCellClass(r, c)).text('');

                cells[r][c] = cell;
                $container.append(cell);
            }
        }
    }

    function getOffset(scroll, length) {
        return Math.floor(scroll / length);
    }

    function getNodeValue(r, c) {
        var rowValue = (r + rowOffset);
        return rowValue + ',' + (c + colOffset);
    }

    function writeTableValues() {
        for (var r = 0; r < rows; r++) {
            for (var c = 0; c < cols; c++) {
                var $cell = cells[r][c];
                if (useTextNodes) {
                    $cell[0].childNodes[0].nodeValue = getNodeValue(r, c);
                } else {
                    $cell[0].innerHTML = '<span>' + (r + rowOffset) + ',' + (c + colOffset) + '</span>';
                }
            }
        }
        if (rowOffset % 2 === 1) {
            $container.addClass('odds');
        } else {
            $container.removeClass('odds');
        }
    }

    var calcOffsetAndDraw = function () {
        var newRowOffset = getOffset(scrollTop, cellHeight);
        var newColOffset = getOffset(scrollLeft, cellWidth);
        if (newColOffset !== colOffset || newRowOffset !== rowOffset) {
            colOffset = newColOffset;
            rowOffset = newRowOffset;
            writeTableValues();
        }
    }.debounce(1);

    //scroll
    var onMouseWheel = function (e) {
        e.preventDefault();
        var event = e.originalEvent;
        var yDelta = mousewheel.getDelta(event);
        var xDelta = mousewheel.getDelta(event, true);
        scrollTop = rowRange.clamp(scrollTop - yDelta);
        scrollLeft = colRange.clamp(scrollLeft - xDelta);
        //console.log('delta: ', xDelta, yDelta);
        //console.log('scroll: ', scrollLeft, scrollTop);
        calcOffsetAndDraw();
    };
    $container.on('mousewheel', onMouseWheel);
    $container.on('wheel', onMouseWheel);

    return {
        setDimensions: function (r, c) {
            maxRows = r;
            maxCols = c;
        },
        rebuild: function () {
            buildCells();
            writeTableValues();
        },
        setUseTextNodes: function (use) {
            useTextNodes = use;
        }
    };
};
